### PR TITLE
BUG: Fix misleading assert message in assert_almost_equal #12200

### LINKED ISSUE #12200: Misleading assert_almost_equal error message in presence of infs.
When using np.testing.assert_almost_equal (and related functions like assert_allclose) to compare arrays which include infs, the error message can be misleading. It describes a comparison between arrays with the infs stripped out, which makes debugging more difficult.

### Reproducing code example:

```python
import numpy as np
np.testing.assert_almost_equal([np.inf, 0], [np.inf, 1])
```

In numpy version 1.12.1, this produces the error message 
```
AssertionError: 
Arrays are not almost equal to 7 decimals

(mismatch 100.0%)
 x: array([ inf,   0.])
 y: array([ inf,   1.])
```

In numpy version 1.13.3, we instead get the error message
```
AssertionError: 
Arrays are not almost equal to 7 decimals

(mismatch 100.0%)
 x: array([ 0.])
 y: array([ 1.])
```

This second error message doesn't reflect what arrays are actually being compared. (The mismatch percentage is also wrong, in both versions.)

### PR DESCRIPTION
**Pull request by liwt31**
Fixes #12200 by making a copy of the matrix before NaN's are excluded

EDIT: Code to test the new error message:
```python
a = np.array([np.inf, 0])
b = np.array([np.inf, 1])
np.testing.assert_almost_equal(a, b)
```
Output:
```
AssertionError: 
Arrays are not almost equal to 7 decimals

(mismatch 50.0%)
 x: array([inf,  0.])
 y: array([inf,  1.])
```

### CODE REVIEW

**Comment by liwt31** [comment]
@tylerjereddy I believe the Travis-CI check has already passed, though the status in this PR page is not updated.

**Comment by mattip** [comment]
Can you add the code you used to test the fix to `TestBuildErrorMessage` in `testing/tests/test_utils.py'.

File: numpy/testing/_private/utils.py [hunk 1]
```diff
@@ -780,10 +783,10 @@ def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
         # do not trigger a failure (np.ma.masked != True evaluates as
         # np.ma.masked, which is falsy).
         if cond != True:
-            match = 100-100.0*reduced.count(1)/len(reduced)
-            msg = build_err_msg([x, y],
+            mismatch = 100.0 * reduced.count(0) / ox.size
+            msg = build_err_msg([ox, oy],
                                 err_msg
-                                + '\n(mismatch %s%%)' % (match,),
+                                + '\n(mismatch %s%%)' % (mismatch,),
```
Comment (mattip) [inline numpy/testing/_private/utils.py]: Why did you change the calculation here? Can you show this is more correct?

Comment (liwt31) [reply numpy/testing/_private/utils.py]: Sure. First of all, the variable name `match` is misleading, because what it actually means is `mismatch`, which you can tell from the string the variable tries to format.
The original way to calculate `match` (which is actually mismatch) is firstly calculate match percentage and then subtract it from 100. This logic is verbose and not compatible with this bug fix, because in this bug fix we are trying to count NaN's as matched result, which is not taken into account in the NaN-free `reduced`. So here it's not feasible to count results that matched (`1`), however results that are not matched (`0`) are stil meaningful here.

**Comment by liwt31** [comment]
Test added and commit squashed. Adding test in `TestBuildErrorMessage` seems not necessary because I didn't change anything in `build_err_msg`. What I did is changing the parameter when calling it. So I suppose add test in `TestAlmostEqual` might be a better idea.

**Comment by mattip** [comment]
Thanks @liwt31
